torch/_dynamo/variables/torch.py

```diff
@@ -804,50 +804,88 @@ def speculate_subgraph(
+            )
+
+        return new_arg
+
+    def add_kwarg_to_subgraph(a):
+        assert not isinstance(a, torch.Tensor), "Tensors should already be tracked?"
+        if a is None:
+            a = ConstantVariable(None)
+
+        if isinstance(a, ConstantVariable):
+            # This arg is not used in the body of the higher order op.
+            # Currently, this new input is added to make the calls
+            # happy, which expect a fixed number of arguments. In
+            # future, we can clean this up.
+            new_proxy = tracer.create_graph_input("const")
+            # Ensures that we recompile when the constant value changes
+            a.add_guard(GuardBuilder.CONSTANT_MATCH)
+            new_arg = a
+        elif isinstance(a, TensorVariable):
+            new_proxy = tracer.create_graph_input(a.as_proxy().node.name)
+            example_value = a.as_proxy().node.meta["example_value"]
+            new_arg = wrap_fx_proxy(tx=tx, proxy=new_proxy, example_value=example_value)
+        else:
+            raise unimplemented(
+                "HigherOrderOperator with body that accepts non-Tensors as input"
+            )
+
+        name = new_proxy.node.name
+        return new_arg, name
+
     try:
         with tx.output.new_subtracer() as tracer:
             args = []
+            kwargs = {}
             # One argument to graph per sub_args
             for a in sub_args:
-                assert not isinstance(
-                    a, torch.Tensor
-                ), "Tensors should already be tracked?"
-                if a is None:
-                    a = ConstantVariable(None)
-
-                if isinstance(a, ConstantVariable):
-                    # This arg is not used in the body of the higher order op.
-                    # Currently, this new input is added to make the calls
-                    # happy, which expect a fixed number of arguments. In
-                    # future, we can clean this up.
-                    tracer.create_graph_input("const")
-                    # Ensures that we recompile when the constant value changes
-                    a.add_guard(GuardBuilder.CONSTANT_MATCH)
-                    new_arg = a
-                elif isinstance(a, TensorVariable):
-                    new_proxy = tracer.create_graph_input(a.as_proxy().node.name)
-                    example_value = a.as_proxy().node.meta["example_value"]
-                    new_arg = wrap_fx_proxy(
-                        tx=tx, proxy=new_proxy, example_value=example_value
-                    )
-                elif isinstance(a, AutogradFunctionContextVariable):
-                    tracer.create_graph_input(a.as_proxy().node.name)
-                    new_arg = a
-                else:
-                    raise unimplemented(
-                        "HigherOrderOperator with body that accepts non-Tensors as input"
-                    )
+                new_arg = add_arg_to_subgraph(a)
```

Comment:
Let's say we're calling speculate_subgraph(f, sub_args, sub_kwargs). The desired handling should be something like:
- We start a new output subgraph
- For every Tensor in `sub_args, sub_kwargs`, we record it as an input to the graph, and then turn that into a new VariableTracker. This gives us `flat_inputs`. Ideally we do not record constants as inputs to the graph (we do that right now out of convenience), though we may need to propagate guards for them.
- We take `flat_inputs`, and then massage it into (args, kwargs) that can then be passed to `call_function`.

It's probably worth experimenting on the `wrap` HigherOrderOp first (i.e. extend wrap to support kwargs) before trying out `torch.func.grad`, because `wrap` is easier to understand.

The semantics of `wrap(f, *args, **kwargs)` are that it invokes `f(*args, **kwargs)`. When Dynamo sees `wrap(f, *args, **kwargs)`, then it should rewrite it into `wrap(gm, *new_args)`, where `gm` is a GraphModule version of `f` that doesn't necessarily take the same args and `new_args` includes the flattened `kwargs`.